File: backend/app/agent/model_client.py
# -*- coding: utf-8 -*-
"""OpenAI-compatible模型客户端，不绑定具体模型供应商。"""
from __future__ import annotations

import json
import os
from dataclasses import dataclass
from typing import Any, AsyncIterator

import httpx

from .model_config import load_local_config


class ModelError(Exception):
    """模型调用失败。"""


class ModelNotConfigured(ModelError):
    """缺少模型地址、模型名或 API Key。"""


@dataclass(frozen=True)
class ModelConfig:
    api_key: str
    base_url: str
    model: str
    timeout_seconds: float = 45.0
    thinking: str = 'disabled'

    @classmethod
    def from_env(cls) -> 'ModelConfig':
        local = load_local_config()
        api_key = os.environ.get('MEIMEI_MODEL_API_KEY') or os.environ.get('OPENAI_API_KEY') or str(local.get('api_key') or '')
        base_url = (os.environ.get('MEIMEI_MODEL_BASE_URL') or str(local.get('base_url') or 'https://api.openai.com/v1')).rstrip('/')
        model = (os.environ.get('MEIMEI_MODEL_NAME') or str(local.get('model') or '')).strip()
        thinking = (os.environ.get('MEIMEI_MODEL_THINKING') or str(local.get('thinking') or 'disabled')).strip().lower()
        if thinking not in {'disabled', 'enabled'}:
            thinking = 'disabled'
        timeout_text = os.environ.get('MEIMEI_MODEL_TIMEOUT', '45')
        try:
            timeout = max(5.0, min(float(timeout_text), 180.0))
        except ValueError:
            timeout = 45.0
        return cls(api_key=api_key, base_url=base_url, model=model, timeout_seconds=timeout, thinking=thinking)

    @property
    def configured(self) -> bool:
        return bool(self.api_key and self.model and self.base_url)


@dataclass(frozen=True)
class ToolCall:
    id: str
    name: str
    arguments: str


@dataclass(frozen=True)
class ModelResponse:
    content: str
    tool_calls: list[ToolCall]
    reasoning_content: str = ''


@dataclass(frozen=True)
class ModelStreamEvent:
    content: str = ''
    response: ModelResponse | None = None


class OpenAICompatibleClient:
    def __init__(self, config: ModelConfig | None = None, http_client: httpx.AsyncClient | None = None):
        self.config = config or ModelConfig.from_env()
        self._http_client = http_client

    async def complete(
        self,
        messages: list[dict[str, Any]],
        tools: list[dict[str, Any]] | None = None,
    ) -> ModelResponse:
        if not self.config.configured:
            raise ModelNotConfigured(
                '模型尚未配置，请设置 MEIMEI_MODEL_API_KEY 和 MEIMEI_MODEL_NAME'
            )
        payload: dict[str, Any] = {
            'model': self.config.model,
            'messages': messages,
            'temperature': 0.2,
            'thinking': {'type': self.config.thinking},
        }
        if tools:
            payload['tools'] = tools
            payload['tool_choice'] = 'auto'

        headers = {
            'Authorization': f'Bearer {self.config.api_key}',
            'Content-Type': 'application/json',
        }
        client = self._http_client
        owns_client = client is None
        if owns_client:
            client = httpx.AsyncClient(timeout=self.config.timeout_seconds)
        try:
            response = await client.post(
                f'{self.config.base_url}/chat/completions',
                headers=headers,
                json=payload,
            )
            if response.status_code >= 400:
                raise ModelError(f'模型接口返回 HTTP {response.status_code}: {_error_text(response)}')
            try:
                data = response.json()
            except ValueError as exc:
                raise ModelError('模型接口返回了无效 JSON') from exc
            return _parse_response(data)
        except httpx.HTTPError as exc:
            raise ModelError(f'模型网络请求失败：{exc}') from exc
        finally:
            if owns_client:
                await client.aclose()

    async def iter_complete(
        self,
        messages: list[dict[str, Any]],
        tools: list[dict[str, Any]] | None = None,
    ) -> AsyncIterator[ModelStreamEvent]:
        """以 OpenAI-compatible SSE 方式流式返回模型文本和最终工具调用。"""
        if not self.config.configured:
            raise ModelNotConfigured(
                '模型尚未配置，请设置 MEIMEI_MODEL_API_KEY 和 MEIMEI_MODEL_NAME'
            )
        payload: dict[str, Any] = {
            'model': self.config.model,
            'messages': messages,
            'temperature': 0.2,
            'thinking': {'type': self.config.thinking},
            'stream': True,
        }
        if tools:
            payload['tools'] = tools
            payload['tool_choice'] = 'auto'
        headers = {
            'Authorization': f'Bearer {self.config.api_key}',
            'Content-Type': 'application/json',
            'Accept': 'text/event-stream',
        }
        client = self._http_client
        owns_client = client is None
        if owns_client:
            client = httpx.AsyncClient(timeout=self.config.timeout_seconds)
        content_parts: list[str] = []
        reasoning_parts: list[str] = []
        tool_calls: dict[int, dict[str, str]] = {}
        try:
            async with client.stream(
                'POST',
                f'{self.config.base_url}/chat/completions',
                headers=headers,
                json=payload,
            ) as response:
                if response.status_code >= 400:
                    await response.aread()
                    raise ModelError(f'模型接口返回 HTTP {response.status_code}: {_error_text(response)}')
                async for line in response.aiter_lines():
                    if not line.startswith('data:'):
                        continue
                    raw = line[5:].strip()
                    if not raw or raw == '[DONE]':
                        if raw == '[DONE]':
                            break
                        continue
                    try:
                        data = json.loads(raw)
                    except ValueError:
                        continue
                    delta = ((data.get('choices') or [{}])[0].get('delta') or {})
                    content = str(delta.get('content') or '')
                    if content:
                        content_parts.append(content)
                        yield ModelStreamEvent(content=content)
                    reasoning = str(delta.get('reasoning_content') or '')
                    if reasoning:
                        reasoning_parts.append(reasoning)
                    for item in delta.get('tool_calls') or []:
                        index = int(item.get('index', len(tool_calls)))
                        call = tool_calls.setdefault(index, {'id': '', 'name': '', 'arguments': ''})
                        call['id'] += str(item.get('id') or '')
                        function = item.get('function') or {}
                        call['name'] += str(function.get('name') or '')
                        call['arguments'] += str(function.get('arguments') or '')
        except httpx.HTTPError as exc:
            raise ModelError(f'模型网络请求失败：{exc}') from exc
        finally:
            if owns_client:
                await client.aclose()
        yield ModelStreamEvent(response=ModelResponse(
            content=''.join(content_parts),
            tool_calls=[ToolCall(**tool_calls[index]) for index in sorted(tool_calls)],
            reasoning_content=''.join(reasoning_parts),
        ))


def _parse_response(data: dict[str, Any]) -> ModelResponse:
    try:
        message = data['choices'][0]['message']
    except (KeyError, IndexError, TypeError) as exc:
        raise ModelError('模型响应缺少 choices[0].message') from exc
    calls = []
    for item in message.get('tool_calls') or []:
        function = item.get('function') or {}
        calls.append(ToolCall(
            id=str(item.get('id') or ''),
            name=str(function.get('name') or ''),
            arguments=str(function.get('arguments') or '{}'),
        ))
    content = message.get('content') or ''
    reasoning_content = message.get('reasoning_content') or ''
    return ModelResponse(
        content=str(content),
        tool_calls=calls,
        reasoning_content=str(reasoning_content),
    )


def _error_text(response: httpx.Response) -> str:
    try:
        data = response.json()
        return str(data.get('error', data))[:500]
    except ValueError:
        return response.text[:500]

File: server/src/agent/modelClient.ts
import { ModelConfig } from './modelConfig.js';

export class ModelError extends Error {}

export class ModelNotConfigured extends ModelError {}

export interface ToolCall {
  id: string;
  name: string;
  arguments: string;
}

export interface ModelResponse {
  content: string;
  tool_calls: ToolCall[];
  reasoning_content: string;
  usage: Record<string, number> | null;
}

export interface ModelStreamEvent {
  content: string;
  response: ModelResponse | null;
}

const DSML_PREFIX = String.raw`<\s*(?:[|｜]\s*)+DSML\s*(?:[|｜]\s*)+`;
const DSML_CLOSE = String.raw`</\s*(?:[|｜]\s*)+DSML\s*(?:[|｜]\s*)+`;
const DSML_BLOCK_RE = new RegExp(`${DSML_PREFIX}tool_calls\\s*>(.*?)${DSML_CLOSE}tool_calls\\s*>`, 'gisu');
const DSML_INVOKE_RE = new RegExp(
  `${DSML_PREFIX}invoke\\s+name\\s*=\\s*["']([^"']+)["']`
  + `(?:\\s+arguments\\s*=\\s*["']([^"']*)["'])?\\s*>(.*?)${DSML_CLOSE}invoke\\s*>`,
  'gisu',
);
const DSML_FRAGMENT_RE = new RegExp(String.raw`</?\s*(?:[|｜]\s*)+DSML\b.*?>`, 'gisu');

function modelMessages(messages: Array<Record<string, unknown>>): Array<Record<string, unknown>> {
  return messages.map((message) => {
    const clean = { ...message };
    delete clean.attachment;
    delete clean.display_content;
    return clean;
  });
}

export class OpenAICompatibleClient {
  readonly config: ModelConfig;

  constructor(config?: ModelConfig) {
    this.config = config ?? ModelConfig.fromEnv();
  }

  async complete(
    messages: Array<Record<string, unknown>>,
    tools?: Array<Record<string, unknown>>,
  ): Promise<ModelResponse> {
    if (!this.config.configured) {
      throw new ModelNotConfigured('模型尚未配置，请设置 MEIMEI_MODEL_API_KEY 和 MEIMEI_MODEL_NAME');
    }
    const payload: Record<string, unknown> = {
      model: this.config.model,
      messages: modelMessages(messages),
      temperature: this.config.temperature,
    };
    /* thinking 是非标准扩展字段，只有显式启用才发送，保证严格 OpenAI 兼容端点可接入。 */
    if (this.config.thinking === 'enabled') {
      payload['thinking'] = { type: 'enabled' };
    }
    if (tools && tools.length > 0) {
      payload['tools'] = tools;
      payload['tool_choice'] = 'auto';
    }
    const headers: Record<string, string> = {
      'Authorization': `Bearer ${this.config.api_key}`,
      'Content-Type': 'application/json',
    };
    const url = `${this.config.base_url}/chat/completions`;
    for (let attempt = 0; attempt < 2; attempt++) {
      let response: Response;
      try {
        response = await fetch(url, {
          method: 'POST',
          headers,
          body: JSON.stringify(payload),
          signal: AbortSignal.timeout(Math.max(1, Math.round(this.config.timeout_seconds * 1000))),
        });
      } catch (error) {
        if (attempt === 0) {
          await sleep(350);
          continue;
        }
        throw new ModelError(`模型网络请求失败：${error instanceof Error ? error.message : String(error)}`);
      }
      if (response.status >= 400) {
        const message = `模型接口返回 HTTP ${response.status}: ${await errorText(response)}`;
        if (response.status === 408 || response.status === 429 || response.status >= 500) {
          if (attempt === 0) {
            await sleep(350);
            continue;
          }
        }
        throw new ModelError(message);
      }
      let data: unknown;
      try {
        data = await response.json();
      } catch {
        if (attempt === 0) {
          await sleep(150);
          continue;
        }
        throw new ModelError('模型接口返回了无效 JSON');
      }
      return parseResponse(data as Record<string, unknown>);
    }
    throw new ModelError('模型请求失败');
  }

  async *iter_complete(
    messages: Array<Record<string, unknown>>,
    tools?: Array<Record<string, unknown>>,
  ): AsyncGenerator<ModelStreamEvent> {
    if (!this.config.configured) {
      throw new ModelNotConfigured('模型尚未配置，请设置 MEIMEI_MODEL_API_KEY 和 MEIMEI_MODEL_NAME');
    }
    const payload: Record<string, unknown> = {
      model: this.config.model,
      messages: modelMessages(messages),
      temperature: this.config.temperature,
      stream: true,
      stream_options: { include_usage: true },
    };
    if (this.config.thinking === 'enabled') {
      payload['thinking'] = { type: 'enabled' };
    }
    if (tools && tools.length > 0) {
      payload['tools'] = tools;
      payload['tool_choice'] = 'auto';
    }
    const headers: Record<string, string> = {
      'Authorization': `Bearer ${this.config.api_key}`,
      'Content-Type': 'application/json',
      'Accept': 'text/event-stream',
    };
    const url = `${this.config.base_url}/chat/completions`;
    let response: Response;
    try {
      response = await fetch(url, {
        method: 'POST',
        headers,
        body: JSON.stringify(payload),
        signal: AbortSignal.timeout(Math.max(1, Math.round(this.config.timeout_seconds * 1000))),
      });
    } catch (error) {
      throw new ModelError(`模型网络请求失败：${error instanceof Error ? error.message : String(error)}`);
    }
    if (response.status >= 400) {
      const message = `模型接口返回 HTTP ${response.status}: ${await errorText(response)}`;
      throw new ModelError(message);
    }
    const contentParts: string[] = [];
    const reasoningParts: string[] = [];
    const toolCalls = new Map<number, { id: string; name: string; arguments: string }>();
    let usage: Record<string, number> = {};
    let finished = false;
    const handleLine = (line: string): string | null => {
      if (!line.startsWith('data:')) return null;
      const raw = line.slice(5).trim();
      if (!raw) return null;
      if (raw === '[DONE]') {
        finished = true;
        return null;
      }
      let data: Record<string, unknown>;
      try {
        data = JSON.parse(raw) as Record<string, unknown>;
      } catch {
        return null;
      }
      if (typeof data['usage'] === 'object' && data['usage'] !== null) {
        usage = filterUsage(data['usage']) ?? {};
      }
      const choices = (Array.isArray(data['choices']) ? data['choices'] : []) as Array<Record<string, unknown>>;
      const delta = (choices.length > 0 && typeof choices[0]?.['delta'] === 'object' && choices[0]['delta'] !== null
        ? choices[0]['delta'] : {}) as Record<string, unknown>;
      const content = String(delta['content'] ?? '');
      if (content) {
        contentParts.push(content);
        if (!looksLikeDsml(contentParts.join(''))) {
          return content;
        }
      }
      const reasoning = String(delta['reasoning_content'] ?? '');
      if (reasoning) {
        reasoningParts.push(reasoning);
      }
      const deltaCalls = (Array.isArray(delta['tool_calls']) ? delta['tool_calls'] : []) as Array<Record<string, unknown>>;
      for (const item of deltaCalls) {
        const index = item['index'] === undefined ? toolCalls.size : Number(item['index']);
        let call = toolCalls.get(index);
        if (!call) {
          call = { id: '', name: '', arguments: '' };
          toolCalls.set(index, call);
        }
        call.id += String(item['id'] ?? '');
        const fn = (typeof item['function'] === 'object' && item['function'] !== null
          ? item['function'] : {}) as Record<string, unknown>;
        call.name += String(fn['name'] ?? '');
        call.arguments += String(fn['arguments'] ?? '');
      }
      return null;
    };
    if (response.body) {
      const reader = response.body.getReader();
      const decoder = new TextDecoder('utf-8');
      let buffer = '';
      try {
        for (;;) {
          const { done, value } = await reader.read();
          if (done) break;
          buffer += decoder.decode(value, { stream: true });
          let newlineIndex = buffer.indexOf('\n');
          while (newlineIndex >= 0) {
            const line = buffer.slice(0, newlineIndex);
            buffer = buffer.slice(newlineIndex + 1);
            const chunk = handleLine(line);
            if (chunk) {
              yield { content: chunk, response: null };
            }
            if (finished) break;
            newlineIndex = buffer.indexOf('\n');
          }
          if (finished) break;
        }
        if (buffer) {
          const chunk = handleLine(buffer);
          if (chunk) {
            yield { content: chunk, response: null };
          }
        }
      } catch (error) {
        if (error instanceof ModelError) throw error;
        throw new ModelError(`模型网络请求失败：${error instanceof Error ? error.message : String(error)}`);
      }
    }
    const [parsedContent, dsmlCalls] = parseDsmlToolCalls(contentParts.join(''));
    const orderedCalls: ToolCall[] = [];
    for (const index of [...toolCalls.keys()].sort((a, b) => a - b)) {
      const call = toolCalls.get(index)!;
      orderedCalls.push({ id: call.id, name: call.name, arguments: call.arguments });
    }
    yield {
      content: '',
      response: {
        content: parsedContent,
        tool_calls: [...orderedCalls, ...dsmlCalls],
        reasoning_content: reasoningParts.join(''),
        usage: Object.keys(usage).length > 0 ? usage : null,
      },
    };
  }
}

function parseResponse(data: Record<string, unknown>): ModelResponse {
  let message: Record<string, unknown>;
  try {
    const choices = (Array.isArray(data['choices']) ? data['choices'] : []) as Array<Record<string, unknown>>;
    if (choices.length === 0 || typeof choices[0]?.['message'] !== 'object' || choices[0]['message'] === null) {
      throw new Error();
    }
    message = choices[0]['message'] as Record<string, unknown>;
  } catch {
    throw new ModelError('模型响应缺少 choices[0].message');
  }
  const calls: ToolCall[] = [];
  for (const item of (Array.isArray(message['tool_calls']) ? message['tool_calls'] : []) as Array<Record<string, unknown>>) {
    const fn = (typeof item['function'] === 'object' && item['function'] !== null
      ? item['function'] : {}) as Record<string, unknown>;
    calls.push({
      id: String(item['id'] ?? ''),
      name: String(fn['name'] ?? ''),
      arguments: String(fn['arguments'] ?? '{}'),
    });
  }
  const [content, dsmlCalls] = parseDsmlToolCalls(String(message['content'] ?? ''));
  return {
    content,
    tool_calls: [...calls, ...dsmlCalls],
    reasoning_content: String(message['reasoning_content'] ?? ''),
    usage: filterUsage(data['usage']),
  };
}

function filterUsage(usage: unknown): Record<string, number> | null {
  if (typeof usage !== 'object' || usage === null) return null;
  const result: Record<string, number> = {};
  const raw = usage as Record<string, unknown>;
  for (const key of ['prompt_tokens', 'completion_tokens', 'total_tokens']) {
    if (key in raw) {
      const value = Number(raw[key]);
      result[key] = Number.isFinite(value) ? Math.trunc(value) : 0;
    }
  }
  return Object.keys(result).length > 0 ? result : null;
}

async function errorText(response: Response): Promise<string> {
  let body: string;
  try {
    body = await response.text();
  } catch {
    return '';
  }
  let data: unknown;
  try {
    data = JSON.parse(body);
  } catch {
    return body.slice(0, 500);
  }
  const value = (data as Record<string, unknown>)?.['error'] ?? data;
  if (typeof value === 'string') return value.slice(0, 500);
  return JSON.stringify(value).slice(0, 500);
}

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

function looksLikeDsml(content: string): boolean {
  const head = String(content ?? '').trimStart().slice(0, 160);
  return head.includes('<') && head.toUpperCase().includes('DSML');
}

function parseDsmlToolCalls(content: string): [string, ToolCall[]] {
  const text = String(content ?? '');
  const calls: ToolCall[] = [];
  const blocks = [...text.matchAll(DSML_BLOCK_RE)];
  for (const block of blocks) {
    const body = block[1] ?? '';
    const invokes = [...body.matchAll(DSML_INVOKE_RE)];
    for (const invoke of invokes) {
      const argumentsText = invoke[2] || (invoke[3] ?? '').trim() || '{}';
      calls.push({
        id: `dsml-call-${calls.length + 1}`,
        name: (invoke[1] ?? '').trim(),
        arguments: argumentsText,
      });
    }
  }
  let rest = text;
  for (const block of [...blocks].reverse()) {
    const start = block.index ?? 0;
    rest = rest.slice(0, start) + rest.slice(start + block[0].length);
  }
  if (calls.length === 0) {
    const invokes = [...rest.matchAll(DSML_INVOKE_RE)];
    for (let index = 0; index < invokes.length; index++) {
      const invoke = invokes[index];
      const argumentsText = invoke[2] || (invoke[3] ?? '').trim() || '{}';
      calls.push({
        id: `dsml-call-${index + 1}`,
        name: (invoke[1] ?? '').trim(),
        arguments: argumentsText,
      });
    }
    rest = rest.replaceAll(DSML_INVOKE_RE, '');
  }
  rest = rest.replaceAll(DSML_FRAGMENT_RE, '').trim();
  return [rest, calls];
}
